arbor/morph/region.cpp

```diff
@@ -0,0 +1,255 @@
+#include <set>
+#include <string>
+#include <vector>
+
+#include <arbor/morph/error.hpp>
+#include <arbor/morph/locset.hpp>
+#include <arbor/morph/primitives.hpp>
+#include <arbor/morph/region.hpp>
+
+#include "morph/em_morphology.hpp"
+#include "util/span.hpp"
+#include "util/strprintf.hpp"
+#include "util/range.hpp"
+#include "util/rangeutil.hpp"
+
+namespace arb {
+namespace reg {
+
+mcable_list merge(const mcable_list& lhs, const mcable_list& rhs) {
+    mcable_list v;
+    v.resize(lhs.size() + rhs.size());
+    std::merge(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), v.begin());
+    return v;
+}
+
+bool is_disjoint_union(const mcable& a, const mcable& b) {
+    if (a.branch!=b.branch) return true;
+    return a<b? a.dist_pos<b.prox_pos: b.dist_pos<a.prox_pos;
+}
+
+bool is_nonnull_intersection(const mcable& a, const mcable& b) {
+    if (a==b) return true; // handles special case of spherical branch
+    if (a.branch!=b.branch) return false;
+    return a<b? a.dist_pos>b.prox_pos: b.dist_pos>a.prox_pos;
+}
+
+mcable make_union(const mcable& a, const mcable& b) {
+    assert(!is_disjoint_union(a,b));
+    return mcable{a.branch, std::min(a.prox_pos, b.prox_pos), std::max(a.dist_pos, b.dist_pos)};
+}
+
+mcable make_intersection(const mcable& a, const mcable& b) {
+    assert(is_nonnull_intersection(a,b));
+
+    return mcable{a.branch, std::max(a.prox_pos, b.prox_pos), std::min(a.dist_pos, b.dist_pos)};
+}
+
+//
+// Explicit cable section
+//
+
+struct cable_ {
+    mcable cable;
+};
+
+region cable(mcable c) {
+    if (!test_invariants(c)) {
+        throw morphology_error(util::pprintf("Invalid cable section {}", c));
+    }
+    return region(cable_{c});
+}
+
+region branch(msize_t bid) {
+    return cable({bid, 0, 1});
+}
+
+mcable_list concretise_(const cable_& reg, const em_morphology& em) {
+    auto& m = em.morph();
```

Comment:
I'm happy with `concretise` as a name, though it is a bit of a mouthful. If we want to consider something a bit easier, `reify` may be an alternative.

Reply:
It isn't that much of a mouthful! `reify` is a bit obscure: it wasn't in my active vocabulary.

We could just use `concrete`?

Reply:
Or `realize`?

Reply:
`concretise` is fine though.

Reply:
Right, then, `thingify` it is!